src/gbinder_fmq.c

```diff
@@ -0,0 +1,851 @@
+} GBinderFmqPriv;
+
+static inline GBinderFmqPriv* gbinder_fmq_cast(GBinderFmq* client)
+    { return G_CAST(client, GBinderFmqPriv, pub); }
+
+GBinderFmqGrantorDescriptor*
+gbinder_fmq_get_grantor_descriptor(
+    GBinderFmq* self,
+    gint index)
+{
+    return &((GBinderFmqGrantorDescriptor *)(
+        self->desc->grantors.data.ptr))[index];
+}
+
+gsize
+gbinder_fmq_available_to_read_bytes(
+    GBinderFmq* self,
+    gboolean contiguous)
+{
+    GBinderFmqPriv* priv = gbinder_fmq_cast(self);
+
+    guint64 read_ptr =
+        atomic_load_explicit(priv->read_ptr, memory_order_acquire);
+    gsize available_to_read =
+        atomic_load_explicit(priv->write_ptr, memory_order_acquire) - read_ptr;
+
+    if (contiguous) {
+        gsize size =
+            gbinder_fmq_get_grantor_descriptor(self, DATA_PTR_POS)->extent;
+        gsize read_offset = read_ptr % size;
+        /* The number of bytes that can be read contiguously from
+         * read_offset without wrapping around the ring buffer */
+        gsize available_to_read_contiguous = size - read_offset;
+        return available_to_read_contiguous < available_to_read ?
+            available_to_read_contiguous : available_to_read;
+    }
+    return available_to_read;
+}
+
+gsize
+gbinder_fmq_available_to_write_bytes(
+    GBinderFmq* self,
+    gboolean contiguous)
+{
+    GBinderFmqPriv* priv = gbinder_fmq_cast(self);
+
+    gsize available_to_write =
+        gbinder_fmq_get_grantor_descriptor(self, DATA_PTR_POS)->extent -
+        gbinder_fmq_available_to_read_bytes(self, FALSE);
+
+    if (contiguous) {
+        guint64 write_ptr = atomic_load_explicit(priv->write_ptr,
+            memory_order_relaxed);
+        gsize size =
+            gbinder_fmq_get_grantor_descriptor(self, DATA_PTR_POS)->extent;
+        gsize write_offset = write_ptr % size;
+        /* The number of bytes that can be written contiguously starting from
+         * write_offset without wrapping around the ring buffer */
+        gsize available_to_write_contiguous = size - write_offset;
+        return available_to_write_contiguous < available_to_write ?
+            available_to_write_contiguous : available_to_write;
+    }
+
+    return available_to_write;
+}
+
+static inline gsize align_to_word_boundary(gsize length) {
+    const gsize alignment_size = 64;
+
+    GASSERT(alignment_size % __WORDSIZE == 0);
+    GASSERT(length <= G_MAXSIZE - alignment_size / 8 + 1);
+
+    return (length + alignment_size / 8 - 1) & ~(alignment_size / 8 - 1U);
+}
+
+GBinderFmqGrantorDescriptor*
+gbinder_fmq_create_grantors(
+    gsize queue_size_bytes,
+    gsize num_fds,
+    gboolean configure_event_flag)
+{
+    gsize num_grantors = configure_event_flag ? EVENT_FLAG_PTR_POS + 1
+                                              : DATA_PTR_POS + 1;
+
+    GBinderFmqGrantorDescriptor* grantors =
+        (GBinderFmqGrantorDescriptor*)g_malloc(num_grantors *
+            sizeof(GBinderFmqGrantorDescriptor));
```

Comment:
`g_new(GBinderFmqGrantorDescriptor, num_grantors)` would be shorter and safer, and I would consider using `g_new0` to zero-initialize the allocated memory and then not touching those fields which remain zero.

Reply:
Changed to use g_new0.